lib/dynflow/director/execution_plan_manager.rb

```diff
@@ -0,0 +1,109 @@
+module Dynflow
+  class Director
+    class ExecutionPlanManager
+      include Algebrick::TypeCheck
+      include Algebrick::Matching
+
+      attr_reader :execution_plan, :future
+
+      def initialize(world, execution_plan, future)
+        @world                 = Type! world, World
+        @execution_plan        = Type! execution_plan, ExecutionPlan
+        @future                = Type! future, Concurrent::Edge::Future
+        @running_steps_manager = RunningStepsManager.new(world)
+
+        unless [:planned, :paused].include? execution_plan.state
+          raise "execution_plan is not in pending or paused state, it's #{execution_plan.state}"
+        end
+        execution_plan.execution_history.add('start execution', @world.id)
+        execution_plan.update_state(:running)
+      end
+
+      def start
+        raise "The future was already set" if @future.completed?
+        start_run or start_finalize or finish
+      end
+
+      def prepare_next_step(step)
+        StepWorkItem.new(execution_plan.id, step).tap do |work|
+          @running_steps_manager.add(step, work)
+        end
+      end
+
+      # @return [Array<WorkItem>] of Work items to continue with
+      def what_is_next(work)
+        Type! work, WorkItem
+
+        compute_next_from_step =-> step do
```

Comment:
I didn't do too much refactorints inside the methods, to not include too many changes at once, but I can update this one. Agree this is not very idiomatic ruby